deployment controller null pointer exception

<!-- This form is for bug reports and feature requests ONLY! 

If you're looking for help check [Stack Overflow](https://stackoverflow.com/questions/tagged/kubernetes) and the [troubleshooting guide](https://kubernetes.io/docs/tasks/debug-application-cluster/troubleshooting/).
-->

**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind bug

**What happened**:
The controller manager crashed and then could not start up again due to NPE.

**What you expected to happen**:
It to startup or recover from this state automatically.

**How to reproduce it (as minimally and precisely as possible)**:
Unfortunately I haven't been able to reproduce in a separate cluster (yet).  Steps that may have caused it: shortly after we had done approx 300 deployment scale down actions using " kubectl scale --replicas=0". We then attempted to delete approx 100 services (with serviceType:LoadBalancer) that had created ELBs (and therefor needed to delete ELBs).  This scenario was triggered and I noticed because only two ELBs succeeded in deleting before the crash prevented further deletion.  We were not able to recover from the nil pointer but we do still have this cluster in a stuck state for more investigation. 

**Anything else we need to know?**:
Traceback from the controller manager is:
```
I0920 01:10:37.699697       1 controller_utils.go:1001] Caches are synced for GC controller
I0920 01:10:37.700511       1 controller_utils.go:1001] Caches are synced for RC controller
I0920 01:10:37.700690       1 controller_utils.go:1001] Caches are synced for endpoint controller
I0920 01:10:37.701144       1 controller_utils.go:1001] Caches are synced for certificate controller
I0920 01:10:37.716689       1 controller_utils.go:1001] Caches are synced for resource quota controller
I0920 01:10:37.743727       1 controller_utils.go:1001] Caches are synced for attach detach controller
I0920 01:10:38.026325       1 controller_utils.go:1001] Caches are synced for deployment controller
I0920 01:10:38.131808       1 controller_utils.go:1001] Caches are synced for garbage collector controller
I0920 01:10:38.131901       1 garbagecollector.go:132] Garbage collector: all resource monitors have synced. Proceeding to collect garbage
I0920 01:10:38.132769       1 controller_utils.go:1001] Caches are synced for replica set controller
I0920 01:10:38.141202       1 controller_utils.go:1001] Caches are synced for disruption controller
I0920 01:10:38.141249       1 disruption.go:305] Sending events to api server.
I0920 01:10:38.560887       1 event.go:218] Event(v1.ObjectReference{Kind:"ReplicaSet", Namespace:"kube-system", Name:"heapster-1152096127", UID:"079405ef-9da0-11e7-be81-12ee649370f8", APIVersion:"extensions", ResourceVersion:"2341744", FieldPath:""}): type: 'Normal' reason: 'SuccessfulDelete' Deleted pod: heapster-1152096127-61wrv
E0920 01:10:39.782309       1 runtime.go:66] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:72
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
/usr/local/go/src/runtime/asm_amd64.s:514
/usr/local/go/src/runtime/panic.go:489
/usr/local/go/src/runtime/panic.go:63
/usr/local/go/src/runtime/signal_unix.go:290
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/deployment/progress.go:100
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/deployment/recreate.go:49
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/deployment/deployment_controller.go:686
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/deployment/deployment_controller.go:135
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/deployment/deployment_controller.go:471
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/deployment/deployment_controller.go:460
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/deployment/deployment_controller.go:160
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:97
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:98
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:52
/usr/local/go/src/runtime/asm_amd64.s:2197
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x355229c]

goroutine 1075 [running]:
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x126
panic(0x4304920, 0x8f474a0)
	/usr/local/go/src/runtime/panic.go:489 +0x2cf
k8s.io/kubernetes/pkg/controller/deployment.(*DeploymentController).syncRolloutStatus(0xc420add180, 0xc429164380, 0x2, 0x2, 0x0, 0xc42196d800, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/deployment/progress.go:100 +0x9cc
k8s.io/kubernetes/pkg/controller/deployment.(*DeploymentController).rolloutRecreate(0xc420add180, 0xc42196d800, 0xc42806e0a0, 0x1, 0x1, 0xc4291ca060, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/deployment/recreate.go:49 +0x485
k8s.io/kubernetes/pkg/controller/deployment.(*DeploymentController).syncDeployment(0xc420add180, 0xc422090680, 0x32, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/deployment/deployment_controller.go:686 +0x990
k8s.io/kubernetes/pkg/controller/deployment.(*DeploymentController).(k8s.io/kubernetes/pkg/controller/deployment.syncDeployment)-fm(0xc422090680, 0x32, 0xc42119f060, 0x4081260)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/deployment/deployment_controller.go:135 +0x3e
k8s.io/kubernetes/pkg/controller/deployment.(*DeploymentController).processNextWorkItem(0xc420add180, 0x50d800)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/deployment/deployment_controller.go:471 +0xd9
k8s.io/kubernetes/pkg/controller/deployment.(*DeploymentController).worker(0xc420add180)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/deployment/deployment_controller.go:460 +0x2b
k8s.io/kubernetes/pkg/controller/deployment.(*DeploymentController).(k8s.io/kubernetes/pkg/controller/deployment.worker)-fm()
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/deployment/deployment_controller.go:160 +0x2a
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc4267c2420)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:97 +0x5e
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc4267c2420, 0x3b9aca00, 0x0, 0x2110ff01, 0xc4200837a0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:98 +0xbd
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.Until(0xc4267c2420, 0x3b9aca00, 0xc4200837a0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:52 +0x4d
created by k8s.io/kubernetes/pkg/controller/deployment.(*DeploymentController).Run
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/deployment/deployment_controller.go:160 +0x22d
```

**Environment**:
- Kubernetes version (use `kubectl version`):
Server Version: version.Info{Major:"1", Minor:"7", GitVersion:"v1.7.6+coreos.0", GitCommit:"e14ebed378d5380b9936455584fa454daa67f23b", GitTreeState:"clean", BuildDate:"2017-09-14T21:52:44Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}

- Cloud provider or hardware configuration**:
AWS

- OS (e.g. from /etc/os-release):
NAME="Container Linux by CoreOS"
ID=coreos
VERSION=1465.7.0
VERSION_ID=1465.7.0
BUILD_ID=2017-09-05-2043
PRETTY_NAME="Container Linux by CoreOS 1465.7.0 (Ladybug)"

- Kernel (e.g. `uname -a`):
Linux ip-10-100-102-222.ec2.internal 4.12.10-coreos #1 SMP Tue Sep 5 20:29:13 UTC 2017 x86_64 Intel(R) Xeon(R) CPU E5-2686 v4 @ 2.30GHz GenuineIntel GNU/Linux

- Install tools:
kube-aws

- Others:
Etcd version 3.2.7

--- a/pkg/controller/deployment/progress.go
+++ b/pkg/controller/deployment/progress.go
@@ -52,7 +52,10 @@ func (dc *DeploymentController) syncRolloutStatus(allRSs []*extensions.ReplicaSe
 		case util.DeploymentComplete(d, &newStatus):
 			// Update the deployment conditions with a message for the new replica set that
 			// was successfully deployed. If the condition already exists, we ignore this update.
-			msg := fmt.Sprintf("ReplicaSet %q has successfully progressed.", newRS.Name)
+			msg := fmt.Sprintf("Deployment %q has successfully progressed.", d.Name)
+			if newRS != nil {
+				msg = fmt.Sprintf("ReplicaSet %q has successfully progressed.", newRS.Name)
+			}
 			condition := util.NewDeploymentCondition(extensions.DeploymentProgressing, v1.ConditionTrue, util.NewRSAvailableReason, msg)
 			util.SetDeploymentCondition(&newStatus, *condition)
 

